src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -41,60 +39,71 @@ public Product getProduct(Long id) {
         List<Product> result =products.stream()
                 .filter(s-> s.getPrice()!=null && s.getStock()>0)
                 .collect(Collectors.toList());
-
         if (query!=null) {
-            Map<Product,Integer> productsSearch=new TreeMap<>();
-
-            String[] partQuery = query.split(" ");
-
-            result.stream()
-                    .filter(s ->
-                    {
-                        int i = 0;
-                        for (String part : partQuery) {
-                            if (s.getDescription().contains(part))
-                                i++;
-                        }
-                        if (i == 0) {
-                            return false;
-                        } else {
-                            productsSearch.put(s, i);
-                            return true;
-                        }
-                    })
-                    .collect(Collectors.toList());
-
-            productsSearch.entrySet().stream().sorted(Map.Entry.comparingByValue());
-            result.clear();
-            result.addAll(productsSearch.keySet());
+            result=findProductsByQuery(query, result);
         }
+        result=sortProductsList(result,sortField,sortOrder);
+
+        return result;
+    }
+
+    private List<Product> findProductsByQuery(String query, List<Product> result){
+        Map<Product,Integer> productsSearch=new TreeMap<>();
+
+        String[] partQuery = query.split(" ");
+
+        result.stream()
+                .filter(s ->
+                {
+                    int i = 0;
+                    for (String part : partQuery) {
+                        if (s.getDescription().contains(part))
+                            i++;
+                    }
+                    if (i == 0) {
+                        return false;
+                    } else {
+                        productsSearch.put(s, i);
+                        return true;
+                    }
+                })
+                .collect(Collectors.toList());
+
```

Comment:
Please try to utilize java streams power to implement this - something like this:

Count number of coincidences:

ToIntFunction<Product> matchesCount = product -> (int) Arrays.stream(partQuery)
                .filter(product.getDescription()::contains)
                .count();

and then use matchesCount to sort the list accordingly:

products.stream()
    .filter(product -> Arrays.stream(partQuery).anyMatch(...))
    .sorted(Comparator.comparingInt(matchesCount )
    .collect(...)